### PR TITLE
Don't recurse into allocations, use a global table instead

### LINKED ISSUE #49663: ICE: build cache corruption
Sorry, the testcase is not exactly simple, but:

- checkout this repository on this commit : https://github.com/Smithay/wayland-rs/tree/02e81d6893c90f4ebfa93c70991fa3ab1fd12ff7
- `cargo +beta test --all --all-features`
- apply the following diff to the repository

```
diff --git a/wayland-server/src/lib.rs b/wayland-server/src/lib.rs
index 8685ab04..f83909dd 100644
--- a/wayland-server/src/lib.rs
+++ b/wayland-server/src/lib.rs
@@ -1,3 +1,5 @@
+#![warn(missing_docs)]
+
 #[macro_use]
 extern crate bitflags;
 extern crate libc;
```

- re-run `cargo +beta test --all --all-features`
- ICE : 

```
error: internal compiler error: librustc/ty/context.rs:998: tried to cache DefId(20/0:393 ~ wayland_server[52b1]::generated[0]::c_interfaces[0]::wl_surface_interface[0]), but was already existing as AllocId(
    451
)
```

- run `cargo clean`
- re-run `cargo +beta test --all --all-features`
- builds and run fine

### PR DESCRIPTION
**Pull request by oli-obk**
r? @michaelwoerister 

fixes #49663

### CODE REVIEW

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/364492214) ([raw log](https://api.travis-ci.org/v3/job/364492214/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (613484/613484), completed with 4872 local objects.
---
[00:00:47] configure: rust.quiet-tests     := True
---
[00:04:31] tidy error: /checkout/src/librustc/ty/maps/on_disk_cache.rs:633: line longer than 100 chars
[00:04:32] some tidy checks failed
[00:04:32]
[00:04:32]
[00:04:32] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/tidy" "/checkout/src" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "--no-vendor" "--quiet"
[00:04:32] expected success, got: exit code: 1
[00:04:32]
[00:04:32]
[00:04:32] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test src/tools/tidy
[00:04:32] Build completed unsuccessfully in 0:01:46
[00:04:32] make: *** [tidy] Error 1
[00:04:32] Makefile:79: recipe for target 'tidy' failed
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

File: src/librustc/ty/maps/on_disk_cache.rs [hunk 1]
```diff
@@ -277,6 +280,31 @@ impl<'sess> OnDiskCache<'sess> {
                 diagnostics_index
             };
 
+            let interpret_alloc_index = {
```
Comment (oli-obk) [inline src/librustc/ty/maps/on_disk_cache.rs]: This exact code exists twice, once for `on_disk_cache` and once for `rustc_metadata::encoder`. Not sure how to dedup.

**Comment by oli-obk** [comment]
Ok, now this actually fixes #49663

cc @eddyb for e9ebc7f (minimal impl of indirectly referring to statics)

**Comment by eddyb** [comment]
LGTM

File: src/librustc_metadata/schema.rs [hunk 2]
```diff
@@ -207,6 +207,7 @@ pub struct CrateRoot {
     pub impls: LazySeq<TraitImpls>,
     pub exported_symbols: EncodedExportedSymbols,
     pub wasm_custom_sections: LazySeq<DefIndex>,
+    pub interpret_alloc_index: Vec<u32>,
```
Comment (oli-obk) [inline src/librustc_metadata/schema.rs]: I'm seeing ICEs inside the deserialize impl for this field in 4 stage 2 run-pass tests with aux builds. I can't reproduce on stage 1 and Travis also takes it. Maybe there's some stale data where crates are loaded that were compiled without this field?

Also: should this be LazySeq?

Comment (eddyb) [reply src/librustc_metadata/schema.rs]: If you have `ignore_git` set (which is the default now), incompatible metadata won't be detected. I think I had this issue with test auxiliary crates. And yes, please make it a `LazySeq<u32>`.

Comment (oli-obk) [reply src/librustc_metadata/schema.rs]: I changed it into a `LazySeq`, now I'm seeing the failures on stage 1, too. `ignore_git` doesn't change anything

Comment (oli-obk) [reply src/librustc_metadata/schema.rs]: The failure seems to be in the deserialization in the field after `interpret_alloc_index`.

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/365007890) ([raw log](https://api.travis-ci.org/v3/job/365007890/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Resolving deltas: 100% (613671/613671), completed with 4881 local objects.
---
[00:00:42] configure: rust.quiet-tests     := True
---
[00:23:16] thread 'main' panicked at 'assertion failed: last_min_end <= position', librustc_metadata/encoder.rs:268:17
[00:23:16] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:23:17] error: Could not compile `core`.
[00:23:17]
[00:23:17] Caused by:
[00:23:17]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name core libcore/lib.rs --color always --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=64878136ec7adadb -C extra-filename=-64878136ec7adadb --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-std/release/deps` (exit code: 101)
[00:23:17] command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "--release" "--locked" "--color" "always" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/checkout/src/libstd/Cargo.toml" "--message-format" "json"
[00:23:17] expected success, got: exit code: 101
[00:23:17] thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1085:9
[00:23:17] note: Run with `RUST_BACKTRACE=1` for a backtrace.
[00:23:17] travis_fold:end:stage1-std
[00:23:17] travis_time:end:stage1-std:start=1523437960816315660,finish=1523438017192265823,duration=56375950163
[00:23:17] failed to run: /checkout/obj/build/bootstrap/debug/bootstrap build
[00:23:17] Build completed unsuccessfully in 0:18:29
[00:23:17] Makefile:28: recipe for target 'all' failed
[00:23:17] make: *** [all] Error 1
---
$ find $HOME/Library/Logs/DiagnosticReports -type f -name '*.crash' -not -name '*.stage2-*.crash' -not -name 'com.apple.CoreSimulator.CoreSimulatorService-*.crash' -exec printf travis_fold":start:crashlog\n\033[31;1m%s\033[0m\n" {} \; -exec head -750 {} \; -exec echo travis_fold":"end:crashlog \; || true
find: `/home/travis/Library/Logs/DiagnosticReports': No such file or directory
travis_time:end:2912f28a:start=1523438017746254813,finish=1523438017754102010,duration=7847197
travis_fold:end:after_failure.3
travis_fold:start:after_failure.4
travis_time:start:18c7d788
$ dmesg | grep -i kill
[   10.128139] init: failsafe main process (1094) killed by TERM signal
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

File: src/librustc/mir/interpret/mod.rs [hunk 3]
```diff
@@ -166,26 +168,18 @@ pub fn specialized_encode_alloc_id<
...
     } else if let Some(fn_instance) = tcx.interpret_interner.get_fn(alloc_id) {
         trace!("encoding {:?} with {:#?}", alloc_id, fn_instance);
-        FN_DISCRIMINANT.encode(encoder)?;
+        AllocKind::Fn.encode(encoder)?;
         fn_instance.encode(encoder)?;
-    } else if let Some(did) = tcx.interpret_interner.get_corresponding_static_def_id(alloc_id) {
-        // extern "C" statics don't have allocations, just encode its def_id
-        EXTERN_STATIC_DISCRIMINANT.encode(encoder)?;
+    } else if let Some(did) = tcx.interpret_interner.get_static(alloc_id) {
+        // referring to statics doesn't need to know about their allocations, just hash the DefId
```
Comment (michaelwoerister) [inline src/librustc/mir/interpret/mod.rs]: s/hash/encode ?

**Review by michaelwoerister — COMMENTED** [review]
Thanks, @oli-obk! Looks great.

I'm probably overlooking something: Where are the allocations for statics encoded?

File: src/librustc/mir/interpret/mod.rs (hunk 3 above)
Comment (oli-obk) [reply src/librustc/mir/interpret/mod.rs]: ha yea, this is actual copy paste

File: src/librustc/mir/interpret/mod.rs [hunk 4]
```diff
@@ -166,26 +168,18 @@ pub fn specialized_encode_alloc_id<
...
-            .encode(encoder)?;
     } else if let Some(fn_instance) = tcx.interpret_interner.get_fn(alloc_id) {
         trace!("encoding {:?} with {:#?}", alloc_id, fn_instance);
-        FN_DISCRIMINANT.encode(encoder)?;
+        AllocKind::Fn.encode(encoder)?;
         fn_instance.encode(encoder)?;
-    } else if let Some(did) = tcx.interpret_interner.get_corresponding_static_def_id(alloc_id) {
-        // extern "C" statics don't have allocations, just encode its def_id
-        EXTERN_STATIC_DISCRIMINANT.encode(encoder)?;
+    } else if let Some(did) = tcx.interpret_interner.get_static(alloc_id) {
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc/mir/interpret/mod.rs]: I'd assume that the `get_alloc` branch is the most common. Also both `get_alloc` and `get_static` are just `HashMap` lookups.

File: src/librustc/ty/maps/on_disk_cache.rs [hunk 5]
```diff
@@ -815,27 +840,17 @@ impl<'enc, 'a, 'tcx, E> SpecializedEncoder<interpret::AllocId> for CacheEncoder<
...
-            Entry::Vacant(entry) => {
-                // ensure that we don't place any AllocIds at the very beginning
-                // of the metadata file, because that would end up making our indices
-                // not special. It is essentially impossible for that to happen,
-                // but let's make sure
-                assert!(pos >= interpret::SHORTHAND_START);
-                entry.insert(pos);
-                None
-            },
+        let index = if self.interpret_alloc_ids.insert(*alloc_id) {
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc/ty/maps/on_disk_cache.rs]: That could work

**Review by oli-obk — COMMENTED** [review]
The `AllocId` for a static does not point to the actual static's `Allocation` anymore. Instead it works like function pointers, it's just an ID that you can use to get at the `DefId` of the static. To get at the `Allocation` of a static, you need to run the `const_eval` query.

**Comment by michaelwoerister** [comment]
> The AllocId for a static does not point to the actual static's Allocation anymore. Instead it works like function pointers, it's just an ID that you can use to get at the DefId of the static. To get at the Allocation of a static, you need to run the const_eval query.

OK, so the `static` has an `AllocId` with no corresponding `Allocation`? And `const_eval()` gives you a `Value` which might contain a `MemoryPointer` but the `AllocId` in there is not the same as the `AllocId` of the corresponding `static`?

**Comment by oli-obk** [comment]
Exactly. It's a little extreme, but forward compatible to allow statics to refer to the inside of other statics in the future. Also much harder to get wrong, because you can't accidentally get at the `Allocation` of a static.

```rust
static A: (i32, u32) = (42, 43);
static B: &'static u32 = &A.1;
```

**Comment by michaelwoerister** [comment]
OK, that makes sense to me then. It also explains why the ordering in `specialized_encode_alloc_id()` does not matter.

It would be great if you could try to get rid of the `interpret_alloc_ids` set. Otherwise this should be good to go `:)`

**Comment by oli-obk** [comment]
Rebased and addressed all review comments

**Comment by oli-obk** [comment]
I got two "lgtm", so let's get this moving (beta needs this)

@bors r=michaelwoerister

**Comment by pnkfelix** [comment]
Marking as beta-accepted, with some reservations due to the size of the PR. But since there's already a bors run for a slew of beta backports that includes this currently in progress (due to a process failure), and @michaelwoerister says leaving this out will break incremental compilation.... <shrug>

**Pull request merged**